File: examples/ch13/comparison.c
/**
 * comparison.c - C 语言 system/popen 示例
 * 
 * 功能: 演示 C 语言如何调用外部命令和程序
 * 编译: gcc comparison.c -o comparison.exe
 * 用法: comparison.exe
 * 安全提示: 仅在项目目录内操作
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
#include <windows.h>
#define popen _popen
#define pclose _pclose
#endif

/**
 * 演示 system() 函数
 */
void demo_system(void) {
    int exit_code;
    
    printf("==================================================\n");
    printf("  方式 1: system() - 简单执行\n");
    printf("==================================================\n\n");
    
    /* 简单命令 */
    printf("[C] 执行 'echo' 命令:\n");
    system("echo Hello from system()");
    printf("\n");
    
    /* 获取退出码 */
    printf("[C] 执行 'dir /b' 命令:\n");
    exit_code = system("dir /b *.c");
    printf("[C] 退出码: %d\n\n", exit_code);
    
    /* 失败的命令 */
    printf("[C] 执行不存在的命令:\n");
    exit_code = system("nonexistent_command 2>nul");
    printf("[C] 退出码: %d\n\n", exit_code);
}

/**
 * 演示 popen() 管道读取
 */
void demo_popen_read(void) {
    FILE *pipe;
    char buffer[256];
    
    printf("==================================================\n");
    printf("  方式 2: popen() - 管道读取\n");
    printf("==================================================\n\n");
    
    printf("[C] 通过管道读取 'hostname' 输出:\n");
    pipe = popen("hostname", "r");
    if (pipe) {
        if (fgets(buffer, sizeof(buffer), pipe)) {
            /* 移除末尾换行符 */
            buffer[strcspn(buffer, "\n")] = 0;
            printf("  主机名: %s\n", buffer);
        }
        pclose(pipe);
    }
    printf("\n");
    
    /* 读取多行输出 */
    printf("[C] 通过管道读取 'dir /b' 输出:\n");
    pipe = popen("dir /b *.c", "r");
    if (pipe) {
        int count = 0;
        while (fgets(buffer, sizeof(buffer), pipe)) {
            buffer[strcspn(buffer, "\n")] = 0;
            printf("  文件: %s\n", buffer);
            count++;
        }
        printf("  共 %d 个文件\n", count);
        pclose(pipe);
    }
    printf("\n");
}

/**
 * 演示 popen() 管道写入
 */
void demo_popen_write(void) {
    FILE *pipe;
    char buffer[256];
    
    printf("==================================================\n");
    printf("  方式 3: popen() - 管道写入\n");
    printf("==================================================\n\n");
    
    printf("[C] 通过管道写入数据:\n");
    pipe = popen("findstr /i \"hello\"", "w");
    if (pipe) {
        fprintf(pipe, "Hello World\n");
        fprintf(pipe, "Test Line\n");
        fprintf(pipe, "hello again\n");
        fprintf(pipe, "Another test\n");
        pclose(pipe);
    }
    printf("\n");
}

/**
 * 演示多命令管道
 */
void demo_pipe_chain(void) {
    FILE *pipe;
    char buffer[256];
    
    printf("==================================================\n");
    printf("  方式 4: 管道链\n");
    printf("==================================================\n\n");
    
    printf("[C] 执行管道链命令:\n");
    pipe = popen("echo apple & echo banana & echo cherry | sort", "r");
    if (pipe) {
        printf("  排序结果:\n");
        while (fgets(buffer, sizeof(buffer), pipe)) {
            buffer[strcspn(buffer, "\n")] = 0;
            printf("    %s\n", buffer);
        }
        pclose(pipe);
    }
    printf("\n");
}

/**
 * 演示获取命令输出到字符串
 */
void demo_capture_output(void) {
    FILE *pipe;
    char buffer[1024] = {0};
    char *pos = buffer;
    size_t remaining = sizeof(buffer) - 1;
    
    printf("==================================================\n");
    printf("  方式 5: 捕获完整输出\n");
    printf("==================================================\n\n");
    
    printf("[C] 捕获 'systeminfo' 部分输出:\n");
    pipe = popen("systeminfo | findstr /i \"OS\"", "r");
    if (pipe) {
        while (fgets(pos, remaining, pipe) && remaining > 0) {
            size_t len = strlen(pos);
            pos += len;
            remaining -= len;
        }
        pclose(pipe);
        
        /* 输出捕获的内容 */
        printf("  系统信息:\n");
        char *line = strtok(buffer, "\n");
        while (line) {
            printf("    %s\n", line);
            line = strtok(NULL, "\n");
        }
    }
    printf("\n");
}

/**
 * 主函数
 */
int main(void) {
    printf("\n");
    printf("+================================================+\n");
    printf("|        C 语言 system/popen 示例               |\n");
    printf("+================================================+\n\n");
    
    /* 演示各种调用方式 */
    demo_system();
    demo_popen_read();
    demo_popen_write();
    demo_pipe_chain();
    demo_capture_output();
    
    printf("==================================================\n");
    printf("  演示完成\n");
    printf("==================================================\n");
    
    return 0;
}

File: examples/ch13/comparison.lua
#!/usr/bin/env luajit
--[[
comparison.lua - Lua os.execute/io.popen 示例

功能: 演示 Lua 如何调用外部命令和程序
用法: luajit comparison.lua
安全提示: 仅在项目目录内操作
--]]

--[[
演示 os.execute() 函数
--]]
local function demo_os_execute()
    print("==================================================")
    print("  方式 1: os.execute() - 简单执行")
    print("==================================================")
    print()
    
    -- 简单命令
    print("[Lua] 执行 'echo' 命令:")
    os.execute("echo Hello from os.execute")
    print()
    
    -- 获取退出码
    print("[Lua] 执行 'dir /b' 命令:")
    local exit_code = os.execute("dir /b *.lua")
    print("[Lua] 退出码: " .. tostring(exit_code))
    print()
    
    -- 失败的命令
    print("[Lua] 执行不存在的命令:")
    exit_code = os.execute("nonexistent_command 2>nul")
    print("[Lua] 退出码: " .. tostring(exit_code))
    print()
end

--[[
演示 io.popen() 管道读取
--]]
local function demo_popen_read()
    print("==================================================")
    print("  方式 2: io.popen() - 管道读取")
    print("==================================================")
    print()
    
    -- 读取单行输出
    print("[Lua] 读取 'hostname' 输出:")
    local pipe = io.popen("hostname")
    if pipe then
        local output = pipe:read("*l")
        print("  主机名: " .. (output or "未知"))
        pipe:close()
    end
    print()
    
    -- 读取多行输出
    print("[Lua] 读取 'dir /b' 输出:")
    pipe = io.popen("dir /b *.lua")
    if pipe then
        local count = 0
        for line in pipe:lines() do
            print("  文件: " .. line)
            count = count + 1
        end
        print("  共 " .. count .. " 个文件")
        pipe:close()
    end
    print()
end

--[[
演示 io.popen() 管道写入
--]]
local function demo_popen_write()
    print("==================================================")
    print("  方式 3: io.popen() - 管道写入")
    print("==================================================")
    print()
    
    print("[Lua] 通过管道写入数据:")
    local pipe = io.popen("findstr /i \"lua\"", "w")
    if pipe then
        pipe:write("Hello World\n")
        pipe:write("Lua Script\n")
        pipe:write("Test Line\n")
        pipe:write("lua example\n")
        pipe:close()
    end
    print()
end

--[[
演示多命令管道
--]]
local function demo_pipe_chain()
    print("==================================================")
    print("  方式 4: 管道链")
    print("==================================================")
    print()
    
    print("[Lua] 执行管道链命令:")
    local pipe = io.popen("echo apple & echo banana & echo cherry | sort")
    if pipe then
        print("  排序结果:")
        for line in pipe:lines() do
            print("    " .. line)
        end
        pipe:close()
    end
    print()
end

--[[
演示捕获命令输出到表
--]]
local function demo_capture_to_table()
    print("==================================================")
    print("  方式 5: 捕获输出到表")
    print("==================================================")
    print()
    
    print("[Lua] 捕获 'dir /b' 输出到表:")
    local pipe = io.popen("dir /b *.lua")
    local files = {}
    if pipe then
        for line in pipe:lines() do
            table.insert(files, line)
        end
        pipe:close()
    end
    
    print("  文件列表:")
    for i, file in ipairs(files) do
        print(string.format("    %d. %s", i, file))
    end
    print("  共 " .. #files .. " 个文件")
    print()
end

--[[
演示执行并获取退出码
--]]
local function demo_exit_code()
    print("==================================================")
    print("  方式 6: 获取退出码")
    print("==================================================")
    print()
    
    -- 成功的命令
    print("[Lua] 执行成功的命令:")
    local success, exit_type, exit_code = os.execute("echo test >nul")
    print("  成功: " .. tostring(success))
    print("  退出类型: " .. tostring(exit_type))
    print("  退出码: " .. tostring(exit_code))
    print()
    
    -- 失败的命令
    print("[Lua] 执行失败的命令:")
    success, exit_type, exit_code = os.execute("nonexistent_command 2>nul")
    print("  成功: " .. tostring(success))
    print("  退出类型: " .. tostring(exit_type))
    print("  退出码: " .. tostring(exit_code))
    print()
end

--[[
演示执行外部程序
--]]
local function demo_external_program()
    print("==================================================")
    print("  方式 7: 执行外部程序")
    print("==================================================")
    print()
    
    -- 执行 Python 脚本（如果可用）
    print("[Lua] 尝试执行 Python:")
    local pipe = io.popen("python --version 2>&1")
    if pipe then
        local output = pipe:read("*a")
        if output and output ~= "" then
            print("  " .. output:gsub("\n", ""))
        else
            print("  Python 未安装")
        end
        pipe:close()
    end
    print()
    
    -- 执行 Node.js（如果可用）
    print("[Lua] 尝试执行 Node.js:")
    pipe = io.popen("node --version 2>&1")
    if pipe then
        local output = pipe:read("*a")
        if output and output ~= "" then
            print("  " .. output:gsub("\n", ""))
        else
            print("  Node.js 未安装")
        end
        pipe:close()
    end
    print()
end

--[[
主函数
--]]
local function main()
    print()
    print("+================================================+")
    print("|        Lua os.execute/io.popen 示例           |")
    print("+================================================+")
    print()
    
    -- 演示各种调用方式
    demo_os_execute()
    demo_popen_read()
    demo_popen_write()
    demo_pipe_chain()
    demo_capture_to_table()
    demo_exit_code()
    demo_external_program()
    
    print("==================================================")
    print("  演示完成")
    print("==================================================")
    
    os.exit(0)
end

-- 执行主函数
main()
